### PR TITLE
#476 Add Comments modules, initial checkin

### PR DESCRIPTION
**Pull request by justindongqiang**
@thiennn Here is the initial checkin for adding comments modules. Since this is my first checkin for simplcommerce, I am expecting many questions. Please let me know if you have any concerns and comments. This is just an initial checkin and I will add more changes next and meantime please let me know your feedback. Some questions as follows:

- Seems there are many duplicated code between comments and reviews and we should refactory put some shared code in one place, like js and csss
- Will add nested comments, comment's comment
- Not sure what the commenter's name for, perhaps we should drop it.
- Then is it only logged users are able to add comments
- I put comment status there just for future extension. Should we do the same to comments, just like review, pending status first, then admin will approve the comments to be appeared?

Thanks,
Justin

### CODE REVIEW

**Comment by thiennn** [comment]
Thank @justindongqiang  the code look nice. 
I prefer to keep Reviews and Comments module separate, although there are some duplication but I think it worth to keep independently, autonomy and simplicity.

Some clients might prefer to allow their customers enter comment without login. In that case all they need is enter a name.

We should have an option for admin to set whether the comments are approved automatically when they are entered or it need approved by the admin

**Comment by justindongqiang** [comment]
@thiennn thank you and glad to help.
I will continue working on it to include comment approval setting and nested comments.

**Comment by justindongqiang** [comment]
@thiennn please bear with me and I am almost done with the augularjs part. Currently I am busy with helping my daughter's school work and will be having more to work on it time next week.

**Comment by justindongqiang** [comment]
@thiennn I checked in the rest of the changes to comments module. I used $('#reviews,#comments') so that we don't have to duplicated code in jquery code... please review and let me know your thoughts. One remaining thing is comment's reply needed to add to admin so that admin can approval them.

File: src/Modules/SimplCommerce.Module.Comments/SqlScripts.txt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+﻿-- Add a new setting to determine if comments require approval  
+INSERT into [SimplCommerce].[dbo].[Core_AppSetting values] ('Product.IsCommentsRequireApproval', 'true', 'Catalog', 1)
```
Comment (thiennn) [inline src/Modules/SimplCommerce.Module.Comments/SqlScripts.txt]: I think we should use Data Seeding of EF Core

**Comment by thiennn** [comment]
Could you please merge 2 ef migrations into one. You can do that by delete those 2 migration then get back the SimplDbContextModelSnapshot.cs at the revision right before you had added the first migration, then run add migration.

Sorry for inform late. but don't focus too much on the customer side, I mean the comment view component to be include in the product details. I am thinking about convert it to a mini spa to improve user experience. But will do after this PR being merged

File: src/Modules/SimplCommerce.Module.Comments/Models/Comment.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+        public CommentStatus Status { get; set; }
+
+        public DateTimeOffset CreatedOn { get; set; }
+
+        public string EntityTypeId { get; set; }
+
+        public long EntityId { get; set; }
+
+        public IList<Reply> Replies { get; protected set; } = new List<Reply>();
```
Comment (thiennn) [inline src/Modules/SimplCommerce.Module.Comments/Models/Comment.cs]: public IList<Comment> Replies { get; protected set; } = new List<Reply>();

**Comment by justindongqiang** [comment]
@thiennn great feedback and I will make changes based on your comments. I will have more time after Aug 16th when my daughter's school starts, and I look forward to making a lot of progress after that...

**Comment by thiennn** [comment]
I also have 2 daughters, very love them

**Comment by justindongqiang** [comment]
@thiennn please bear with me, house is painting in these couple weeks...

**Comment by justindongqiang** [comment]
@thiennn can you help to resolve this conflicting file?

**Comment by justindongqiang** [comment]
@thiennn can you check if wishlist project is placed outside src\modules?

image

**Comment by justindongqiang** [comment]
@thiennn I think I fixed all the issues of comment module. Please have a look and let me know if you have any questions about this PR. Thanks for your patient.

File: src/Modules/SimplCommerce.Module.Comments/Components/CommentViewComponent.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+        public async Task<IViewComponentResult> InvokeAsync(long entityId, string entityTypeId)
+        {
+            var model = new CommentVm
+            {
+                EntityId = entityId,
+                EntityTypeId = entityTypeId
+            };
+
+            model.Items.Data = await _commentRepository
```
[the remark this answers is not in the document]
Comment (thiennn) [reply src/Modules/SimplCommerce.Module.Comments/Components/CommentViewComponent.cs]: Thank @JwanKhalaf , we will this later

**Comment by thiennn** [comment]
@justindongqiang Thank you very much

File: src/Modules/SimplCommerce.Module.Comments/Components/CommentViewComponent.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (justindongqiang) [reply src/Modules/SimplCommerce.Module.Comments/Components/CommentViewComponent.cs]: Thank @JwanKhalaf , Yes not now... we should be creating a service layer fro these logic later.

**Comment by justindongqiang** [comment]
@thiennn please let me know which story you would like to work with next.

**Pull request merged**